drivers/wasapi/audio_driver_wasapi.cpp

```diff
@@ -732,241 +814,304 @@ void AudioDriverWASAPI::write_sample(WORD format_tag, int bits_per_sample, BYTE
-							// We reopened the output device and samples_in may have resized, so invalidate the current avail_frames
-							avail_frames = 0;
-						}
+					avail_frames -= write_frames;
+					written_frames += write_frames;
+				} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
+					// output_device is not valid anymore, reopen it
+					Error err = finish_output_device();
+					if (err != OK) {
+						ERR_PRINT("WASAPI: finish_output_device error");
 					} else {
-						ERR_PRINT("WASAPI: Get buffer error");
-						ad->exit_thread.set();
+						// We reopened the output device and samples_in may have resized, so invalidate the current avail_frames
+						avail_frames = 0;
 					}
+				} else {
+					ERR_PRINT("WASAPI: Get buffer error");
+					exit_thread.set();
 				}
-			} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
-				invalidated = true;
-			} else {
-				ERR_PRINT("WASAPI: GetCurrentPadding error");
 			}
+		} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
+			invalidated = true;
+		} else {
+			ERR_PRINT("WASAPI: GetCurrentPadding error");
+		}
 
-			if (invalidated) {
-				// output_device is not valid anymore
-				WARN_PRINT("WASAPI: Current output_device invalidated, closing output_device");
+		if (invalidated) {
+			// output_device is not valid anymore
+			WARN_PRINT("WASAPI: Current output_device invalidated, closing output_device");
 
-				Error err = ad->finish_output_device();
-				if (err != OK) {
-					ERR_PRINT("WASAPI: finish_output_device error");
-				}
+			Error err = finish_output_device();
+			if (err != OK) {
+				ERR_PRINT("WASAPI: finish_output_device error");
 			}
+			// Reset avail_frames to force regeneration of audio data with new buffer size
+			avail_frames = 0;
 		}
+	}
+}
+
+// Helper function to process audio input
+void AudioDriverWASAPI::process_audio_input(uint32_t &read_frames) {
+	if (audio_input.active.is_set()) {
+		UINT32 packet_length = 0;
+		BYTE *data;
+		UINT32 num_frames_available;
+		DWORD flags;
+
+		HRESULT hr = audio_input.capture_client->GetNextPacketSize(&packet_length);
+		if (hr == S_OK) {
+			while (packet_length != 0) {
+				hr = audio_input.capture_client->GetBuffer(&data, &num_frames_available, &flags, nullptr, nullptr);
+				if (hr == S_OK) {
+					read_audio_buffer(data, num_frames_available, flags);
 
-		// If we're using the Default output device and it changed finish it so we'll re-init the output device
-		if (ad->audio_output.device_name == "Default" && default_output_device_changed) {
-			Error err = ad->finish_output_device();
+					read_frames += num_frames_available;
+
+					hr = audio_input.capture_client->ReleaseBuffer(num_frames_available);
+					if (hr != S_OK) {
+						ERR_PRINT("WASAPI: Release buffer error");
+						break;
+					}
+				} else if (hr == AUDCLNT_E_DEVICE_INVALIDATED) {
+					// Input device is not valid anymore, reopen it
+					WARN_PRINT("WASAPI: Current input device invalidated, closing input device");
+					Error err = finish_input_device();
```

Comment:
This is part of the new case, from what I can tell (the old code used `ERR_BREAK(hr != S_OK)`). I want to understand the implication of calling finish_input_device() only here and in the similar if / else in the surrounding code.